labs/saw/arithmetic-verification/arithmetic_implementations.c

```diff
@@ -0,0 +1,198 @@
+  
+  result = local_a * local_b;
+  
+  return result;
+}
+
+// multiply_textbook
+//
+// Multiply two 16-bit numbers by treating each as two 8-bit parts and using
+// the "textbook" multiplication algorithm.
+//
+// Suppose we break up 32-bit integers a and be as follows: 
+//     a = [ a_1 ] [ a_0 ] 
+//     b = [ b_1 ] [ b_0 ]
+//
+// Then we compute a * b as:
+//                                         [ a_1 ] [ a_0 ]
+//                                       x [ b_1 ] [ b_0 ]
+//                                   =====================
+//                             [ a_1 * b_0 ] [ a_0 * b_0 ]
+//               [ a_1 * b_1 ] [ a_0 * b_1 ]       0
+//               =========================================
+// [    O_3    ] [    O_2    ] [    O_1    ] [    O_0    ]
+//
+// Where the outputs are the sums of the columns
+//
+uint32_t multiply_textbook(uint16_t a, uint16_t b) {
+  uint16_t a_1 = (0xff00 & a) >> 8;
+  uint16_t a_0 = (0x00ff & a) >> 0;
+  uint16_t b_1 = (0xff00 & b) >> 8;
+  uint16_t b_0 = (0x00ff & b) >> 0;
+  
+  uint32_t z0 = a_0 * b_0;
+  uint32_t z1 = a_1 * b_0;
+  uint32_t z2 = a_0 * b_1;
+  uint32_t z3 = a_1 * b_1;
+  
+  uint32_t result = 0;
+  result += z0;
+  result += z1 << 8;
+  result += z2 << 8;
+  result += z3 << 16;
+  
+  return result;
+}
+
+// multiply_karatsuba
+//
+// Multiply two 32-bit numbers by treating each as two 16-bit parts and using
+// the (one-step) Karatsuba multiplication algorithm.
+//
+// Suppose we break up 32-bit integers a and be as follows: 
+//     a = [ a_1 ] [ a_0 ] 
+//     b = [ b_1 ] [ b_0 ]
+//
+// Then we compute a * b as follows:
+//
+// [ O_2 ] [ O_1 ] [ O_0 ]  where
+//   [ O_2 ] = [ a_1 ] * [ b_1 ]
+//   [ O_0 ] = [ a_0 ] * [ b_0 ]
+//   [ O_1 ] = ([ a_1 ] + [ a_0] ) * ([ b_1 ]  + [ b_0 ]) - [ z0 ] - [ z2 ]
+//
+//   Note that if we reduce the expression defining [ O_1 ] we arrive at:
+//
+//     [ a_1 ] * [ b_0 ] + [ a_0 ] * [ b_1 ] 
+//
+//   which has _two_ multiplications, while the expression defining [ O_1 ] only
+//   uses _one_multiplication.
+//
+// The generalized Karatsuba algorithm for wider inputs takes advantage of this
+// feature and the assumption that multiplication is a slower computation than
+// addition to recursively compute the product of two numbers. Whenever a
+// product appears in the computation above, one could -- in theory -- replace
+// that with a Karatsuba product of its own.
+//
+uint32_t multiply_karatsuba(uint16_t a, uint16_t b) {
+  uint16_t a_1 = (0xff00 & a) >> 8;
+  uint16_t a_0 = (0x00ff & a) >>  0;
+  uint16_t b_1 = (0xff00 & b) >> 8;
+  uint16_t b_0 = (0x00ff & b) >>  0;
+  
+  uint32_t z0 = a_0 * b_0;
+  uint32_t z2 = a_1 * b_1;
+  uint32_t z1 = (a_1 + a_0) * (b_1 + b_0) - z0 - z2;
+  
+  uint32_t result = 0;
+  result += z0;
+  result += z1 << 8;
+  result += z2 << 16;
+  
+  return result;
+}
+
+
+//
+// The main routine, verifies that the algorithms above match for the collection
+//  of arithmetic algorithms above.
```

Comment:
Repetitive use of phrase "algorithms above"

Reply:
Fixed and clarified the intent slightly.